### PR TITLE
MINOR: Follow-up minor improvements/cleanup for KAFKA-3396

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

**Comment by hachikuji** [comment]
Review by @ijuma.

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala [hunk 1]
```diff
@@ -60,7 +55,7 @@ import scala.util.{Failure, Success}
   * SaslTestHarness here directly because it extends ZooKeeperTestHarness, and we
   * would end up with ZooKeeperTestHarness twice.
   */
-trait EndToEndAuthorizationTest extends IntegrationTestHarness with SaslSetup {
+abstract class EndToEndAuthorizationTest extends IntegrationTestHarness with SaslSetup {
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:58]: What is the reason for this change?

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala [hunk 2]
```diff
@@ -224,27 +217,16 @@ trait EndToEndAuthorizationTest extends IntegrationTestHarness with SaslSetup {
...
 
   /**
     * Tests that a producer fails to publish messages when the appropriate ACL
     * isn't set.
     */
-  @Test
+  @Test(expected = classOf[TimeoutException])
   def testNoProduceWithoutDescribeAcl {
-    //Produce records
-    debug("Starting to send records")
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:241]: I'm not a huge fan of debug statements like this, but supposedly they helped others. Is the idea that we have enough debugging statements in the producer and consumer to figure out these things anyway?

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala [hunk 3]
```diff
@@ -253,45 +235,34 @@ trait EndToEndAuthorizationTest extends IntegrationTestHarness with SaslSetup {
...
     }
-    //Produce records
-    debug("Starting to send records")
     try{
       sendRecords(numRecords, tp)
       fail("exception expected")
     } catch {
-      case e: TopicAuthorizationException => //expected
+      case e: TopicAuthorizationException =>
+      assertEquals(Set(topic).asJava, e.unauthorizedTopics())
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala]: Indent.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 4]
```diff
@@ -592,11 +592,14 @@ private void handleListOffsetResponse(Map<TopicPartition, Long> timestampsToSear
...
                 timestampOffsetMap.put(topicPartition, null);
-            } else if (error == Errors.NOT_LEADER_FOR_PARTITION
-                    || error == Errors.UNKNOWN_TOPIC_OR_PARTITION) {
+            } else if (error == Errors.NOT_LEADER_FOR_PARTITION) {
                 log.debug("Attempt to fetch offsets for partition {} failed due to obsolete leadership information, retrying.",
                         topicPartition);
                 future.raise(error);
+            } else if (error == Errors.UNKNOWN_TOPIC_OR_PARTITION) {
+                log.warn("Received unknown topic or partition error in ListOffset request for partition {}. The topic/partition " +
+                        "may not exist or the user may not have Describe access to it", topicPartition);
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:601]: I guess this could be the logs a little verbose, but one would hope that it would be temporary as it would only be until the topic is automatically created, right?

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala [hunk 5]
```diff
@@ -253,45 +235,34 @@ trait EndToEndAuthorizationTest extends IntegrationTestHarness with SaslSetup {
...
-
-    try {
-      consumeRecords(this.consumers.head)
-      fail("exception expected")
-    } catch {
-      case e: JTimeoutException => //expected
-    }
-  } 
+    // this should timeout since the consumer will not be able to fetch any metadata for the topic
+    consumeRecords(this.consumers.head, timeout=3000)
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala]: Nit: we usually have spaces before and after `=`

File: docs/upgrade.html [hunk 6]
```diff
@@ -64,8 +64,11 @@
...
     <li> The new Java Consumer now allows users to search offsets by timestamp on partitions. </li>
-    <li> When using an Authorizer and a user hasn't got <b>Describe</b> authorization on a topic, the broker will no longer return TOPIC_AUTHORIZATION_FAILED errors  
-         but just UNKNOWN_TOPIC_OR_PARTITION errors, to avoid leaking topic names.</li>
+    <li> When using an Authorizer and a user doesn't have <b>Describe</b> authorization on a topic, the broker will no
+         longer return TOPIC_AUTHORIZATION_FAILED errors to requests since this leaks topic names. Instead, the UNKNOWN_TOPIC_OR_PARTITION
+         error code will be returned. This may cause unexpected timeouts or delays when using the producer and consumer since
+         Kafka clients will typically retry silently on unknown topic errors. You should consult the client logs if you
```
Comment (ijuma) [inline docs/upgrade.html]: Maybe `silently` should be replaced by `automatically`? We do log at `WARN` level, so it's not exactly silent. :)

**Review by ijuma — CHANGES_REQUESTED** [review]
Thanks for the follow-up PR. Overall looks good. Left a few minor comments and I left a few others in the original commit:

https://github.com/apache/kafka/commit/8124f6e0996cb673760750b3aba004ae11e34c6a

Do you think you'd be able to address those too? Some of them are trivial, others a bit more work. We could also just file a JIRA for the latter if easier.

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala (hunk 1 above)
Comment (hachikuji) [reply core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:58]: I found this was required in order to get JUnit to see the `expected` argument in the `@Test` annotation.

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala (hunk 2 above)
Comment (hachikuji) [reply core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:241]: But I guess you can add them back when you need them easily enough? I found them a little distracting when reviewing 3396 and some of them were even incorrect or misleading (which made me think they had probably just been copied from an existing test).

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala (hunk 1 above)
Comment (ijuma) [reply core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:58]: Interesting. Sounds good then.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 4 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:601]: Yeah, I was torn on some of these. I'm fine bumping them down to DEBUG level, but I thought we should make some effort to make the problem more visible since this could now be a persistent problem (if the user does not have Describe access).

File: docs/upgrade.html (hunk 6 above)
Comment (hachikuji) [reply docs/upgrade.html]: Ack

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 7]
```diff
@@ -561,9 +562,16 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
     val consumer = TestUtils.createNewConsumer(TestUtils.getBrokerListStrFromServers(servers), groupId = group,
       securityProtocol = SecurityProtocol.PLAINTEXT, props = Some(consumerConfig))
     try {
+      // ensure that internal topics are not included if no permission
       consumer.subscribe(Pattern.compile(".*"), new NoOpConsumerRebalanceListener)
       consumeRecords(consumer)
       assertEquals(Set(topic).asJava, consumer.subscription)
+
+      // now authorize the user for the internal topic and verify that we can subscribe
+      addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Read)),  new Resource(Topic, kafka.common.Topic.GroupMetadataTopicName))
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: One extra space before `new Resource`.

Comment (ijuma) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Also, `new` is not needed.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 8]
```diff
@@ -561,9 +562,16 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
...
+      consumer.subscribe(Pattern.compile(kafka.common.Topic.GroupMetadataTopicName), new NoOpConsumerRebalanceListener)
+      consumer.poll(0)
+      assertEquals(Set(kafka.common.Topic.GroupMetadataTopicName), consumer.subscription().asScala)
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: `()` not needed in `subscription`

**Comment by ijuma** [comment]
LGTM provided that the tests pass. Left a couple of trivial comments that don't require re-review.
